app/townsquare/views.py

```diff
@@ -0,0 +1,264 @@
+        if request.POST['direction'] == 'liked':
+            Like.objects.create(profile=request.user.profile, activity=activity)
+        if request.POST['direction'] == 'unliked':
+            activity.likes.filter(profile=request.user.profile).delete()
+
+    # flag request
+    elif request.POST.get('method') == 'flag':
+        if request.POST['direction'] == 'flagged':
+            Flag.objects.create(profile=request.user.profile, activity=activity)
+        if request.POST['direction'] == 'unflagged':
+            activity.flags.filter(profile=request.user.profile).delete()
+
+    # comment request
+    elif request.POST.get('method') == 'comment':
+        comment = request.POST.get('comment')
+        comment = Comment.objects.create(profile=request.user.profile, activity=activity, comment=comment)
+        to_emails = set(activity.comments.exclude(profile=request.user.profile).values_list('profile__email', flat=True))
+        comment_email(comment, to_emails)
+
+    elif request.GET.get('method') == 'comment':
+        comments = activity.comments.order_by('-created_on')
+        comments = [comment.to_standard_dict(properties=['profile_handle']) for comment in comments]
+        response['comments'] = comments
+    return JsonResponse(response)
+
+is_debugging_offers = settings.DEBUG
+
+def offer_go(request, offer_id, offer_slug):
+
+    try:
+        offer = Offer.objects.current().get(pk=offer_id)
+        if not request.user.is_authenticated:
+            return redirect('/login/github?next=' + request.get_full_path())
+        if request.user.profile.offeractions.filter(what='go', offer=offer) and not is_debugging_offers:
+            raise Exception('already visited this offer')
+        OfferAction.objects.create(profile=request.user.profile, offer=offer, what='go')
+        return redirect(offer.url)
+    except:
+        raise Http404
+
+
+def offer_decline(request, offer_id, offer_slug):
+
+    try:
+        offer = Offer.objects.current().get(pk=offer_id)
+        if not request.user.is_authenticated:
+            return redirect('/login/github?next=' + request.get_full_path())
+        if request.user.profile.offeractions.filter(what='decline', offer=offer) and not is_debugging_offers:
+            raise Exception('already visited this offer')
+        OfferAction.objects.create(profile=request.user.profile, offer=offer, what='decline')
+        return redirect('/')
+    except:
+        raise Http404
+
+
+def offer_view(request, offer_id, offer_slug):
+
+    try:
+        offer = Offer.objects.current().get(pk=offer_id)
+        if not request.user.is_authenticated:
+            return redirect('/login/github?next=' + request.get_full_path())
+        if request.user.profile.offeractions.filter(what='click', offer=offer) and not is_debugging_offers:
+            raise Exception('already visited this offer')
+        OfferAction.objects.create(profile=request.user.profile, offer=offer, what='click')
```

Comment:
Lots of code in these methods we could DRY up

Reply:
id be open to discussions about how to do that... if we were gonna get fancy we could use an inherited class or view or something..

Reply:
I think just a non-view helper function is probably fine, something like `offer = get_offer_and_create_offer_action(offer_id, what='click')`

Reply:
updated